plugins/metrics/pkg/cortex/intercept.go

```diff
@@ -0,0 +1,192 @@
+package cortex
+
+import (
+	"context"
+
+	"errors"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/weaveworks/common/user"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+const lowerOrgIDHeaderName = "x-scope-orgid"
+
+type WriteHandlerFunc = func(context.Context, *cortexpb.WriteRequest, ...grpc.CallOption) (*cortexpb.WriteResponse, error)
+
+type RequestInterceptor interface {
+	Intercept(ctx context.Context, req *cortexpb.WriteRequest, handler WriteHandlerFunc) (*cortexpb.WriteResponse, error)
+}
+
+type federatingInterceptor struct {
+	conf InterceptorConfig
+
+	mdTracker *metadataTracker
+}
+
+type InterceptorConfig struct {
+	// The label to use as the tenant ID
+	IdLabelName string
+	// Whether to drop the tenant ID label in the output
+	DropIdLabel bool
+}
+
+func NewFederatingInterceptor(conf InterceptorConfig) RequestInterceptor {
+	return federatingInterceptor{
+		conf:      conf,
+		mdTracker: newMetadataTracker(),
+	}
+}
+
+// Intercept takes a write request and splits it into multiple requests, grouped
+// by tenant id found within the configured label name.
+// The outgoingCtx must have an org id set, as well as outgoing metadata, to
+// use as the default tenant id for timeseries without the configured label.
+func (t federatingInterceptor) Intercept(
+	outgoingCtx context.Context, req *cortexpb.WriteRequest,
+	handler WriteHandlerFunc,
+) (*cortexpb.WriteResponse, error) {
+	if len(req.Metadata) > 0 {
+		// Metadata comes in periodically, separate from timeseries, and in one
+		// large batch. Because we are splitting timeseries samples to other
+		// tenants, we also need to replicate the metadata to those tenants.
+		// We need to keep track of names of metrics for which we have sent
+		// timeseries data to, so that we can send metadata for those metrics
+		// to the same tenants the next time metadata comes in.
+		var errs []error
+		_, err := handler(outgoingCtx, req)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		errs = append(errs, t.mdTracker.replicate(req.Metadata, func(tenantId string, metricMetadata []*cortexpb.MetricMetadata) error {
+			req.Metadata = metricMetadata
+			md, _, _ := metadata.FromOutgoingContextRaw(outgoingCtx)
+			md[lowerOrgIDHeaderName] = []string{tenantId}
+			if _, err := handler(outgoingCtx, req); err != nil {
+				return err
+			}
+			return nil
+		})...)
+		return nil, errors.Join(errs...)
+	}
+	// for each time series in the request, check if there is a matching tenant ID label.
+	// if there is, start a new request for this and all subsequent entries with that tenant ID.
+	defaultId, err := user.ExtractOrgID(outgoingCtx)
+	if err != nil {
+		panic(err)
+	}
+	var buckets = map[string][]cortexpb.PreallocTimeseries{
+		defaultId: {},
+	}
+TIMESERIES:
+	for _, ts := range req.Timeseries {
+		nameLabelIndex := -1
```

Comment:
can simplify this block using [lo.FindIdexOf](https://pkg.go.dev/github.com/samber/lo#readme-findindexof)